Framework/DataHandling/src/LoadFITS.cpp

```diff
@@ -451,6 +297,207 @@ void LoadFITS::doLoadHeaders(const std::vector<std::string> &paths,
+        boost::lexical_cast<std::string>(bytes) + " bytes.");
+  }
+
+  // Create a group for these new workspaces, if the group already exists, add
+  // to it.
+  std::string groupName = outWSName;
+
+  size_t fileNumberInGroup = 0;
+  WorkspaceGroup_sptr wsGroup;
+
+  if (!AnalysisDataService::Instance().doesExist(groupName)) {
+    wsGroup = WorkspaceGroup_sptr(new WorkspaceGroup());
+    wsGroup->setTitle(groupName);
+  } else {
+    // Get the name of the latest file in group to start numbering from
+    if (AnalysisDataService::Instance().doesExist(groupName))
+      wsGroup =
+          AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(groupName);
+
+    std::string latestName = wsGroup->getNames().back();
+    // Set next file number
+    fileNumberInGroup = fetchNumber(latestName) + 1;
+  }
+
+  size_t totalWS = headers.size();
+  // Create a progress reporting object
+  API::Progress progress(this, 0, 1, totalWS + 1);
+  progress.report(0, "Loading file(s) into workspace(s)");
+
+  // Create first workspace (with instrument definition). This is also used as
+  // a template for creating others
+  Workspace2D_sptr latestWS;
+  latestWS =
+      makeWorkspace(headers[0], fileNumberInGroup, buffer, imageY, imageE,
+                    latestWS, loadAsRectImg, binSize, noiseThresh);
+  progress.report(1, "First file loaded.");
+
+  std::vector<Workspace2D_sptr> wsOrdered(totalWS);
+  wsOrdered[0] = latestWS;
+
+  if (isInstrOtherThanIMAT(headers[0])) {
+    // For now we assume IMAT except when specific headers are found by
+    // isInstrOtherThanIMAT()
+    //
+    // TODO: do this conditional on INSTR='IMAT' when we have proper IMAT .fits
+    // files
+    try {
+      IAlgorithm_sptr loadInst = createChildAlgorithm("LoadInstrument");
+      std::string directoryName =
+          Kernel::ConfigService::Instance().getInstrumentDirectory();
+      directoryName = directoryName + "/IMAT_Definition.xml";
+      loadInst->setPropertyValue("Filename", directoryName);
+      loadInst->setProperty<MatrixWorkspace_sptr>(
+          "Workspace", boost::dynamic_pointer_cast<MatrixWorkspace>(latestWS));
+      loadInst->execute();
+    } catch (std::exception &ex) {
+      g_log.information("Cannot load the instrument definition. " +
+                        std::string(ex.what()));
+    }
+  }
+
+  // don't feel tempted to parallelize this loop as it is - it uses the same
+  // imageY and imageE buffers for all the workspaces
+  for (int64_t i = 1; i < static_cast<int64_t>(totalWS); ++i) {
+    latestWS =
+        makeWorkspace(headers[i], fileNumberInGroup, buffer, imageY, imageE,
+                      latestWS, loadAsRectImg, binSize, noiseThresh);
+    wsOrdered[i] = latestWS;
+    progress.report("Loaded file " + boost::lexical_cast<std::string>(i + 1) +
+                    " of " + boost::lexical_cast<std::string>(totalWS));
+  }
+
+  // Add to group - done here to maintain sequence
```

Comment:
If I understand correctly, this should not be needed anymore with the parallel for loop gone, or am I overlooking something?